resource/policies/dfu_match_policy_factory.cpp

```
@@ -13,59 +13,167 @@ extern "C" {
+     {"high", "policy=high"},
+     {"low", "policy=low"},
+     {"lonode", "policy=low node_centric=true"},
+     {"hinode", "policy=high node_centric=true"},
+     {"lonodex", "policy=low node_centric=true node_exclusive=true"},
+     {"hinodex", "policy=high node_centric=true node_exclusive=true"},
+     {"locality", ""},
+     {"variation", ""}};
+
+const std::vector<std::string> policy_options = {"policy",
+                                                 "node_centric",
+                                                 "stop_on_1_matches",
+                                                 "node_exclusive"};
+
+/* This takes a string of match policy options, deliniated by
+ * spaces and equals signs, and turns them into a std::map.
+ * It validates that each option is set to either "true" or "false"
+ * and puts the corresponding boolean in the map.
+ */
+bool parse_custom_match_policy (const std::string long_string, std::map<std::string, bool> &split)
+{
+    std::vector<std::string> options;
+    boost::split (options, long_string, boost::is_any_of (" "));
+    /* After splitting based on term, validate the terms. If invalid,
+     * abort.
+     */
+    for (int i = 0; i < options.size (); i++) {
+        std::vector<std::string> settings;
+        /* This split should take an entire string and split it
+         * on the equals sign. If there's no equals sign, maybe it will
+         * get caught later?
+         */
+        boost::split (settings, options.at (i), boost::is_any_of ("="));
+        /* Place the terms in a map for ease of access. Lots of gross
+         * optionsorary structures needed here, hopefully they'll get
+         * garbage collected soon.
+         */
+        if (std::find (policy_options.begin (), policy_options.end (), settings.at (0))
+            == policy_options.end ()) {
+            std::cerr << "invalid policy option: " << settings.at (0) << std::endl;
```

Actually, you should probably avoid writing directly to stdout/err from a broker module.

If there's a way to throw an exception and catch the error message at a higher level, which would then `flux_log (LOG_ERR, ...)` the result, that might be good. Otherwise you'd have to pass down a `flux_t *` handle to this function so it can log errors.

You could also return an error via a separate parameter when the result is `false`. Some of the other Fluxion developers may have a preference here.

---

We had a chat about this yesterday after the meeting. Just above this in the call chain we're passing a std::string through by reference and appending messages to it as necessary, then returning it up to where we can handle it.  The trick is we can't always have a handle here, because this is used in both the module and in `resource_query`.  Essentially the separate parameter option.  Example of the plan on line 516 in resource_match_opts.

I have a separate issue about doing a better job of this in fluxion as a whole, but don't want it to block this PR, so plan is to pass that output string down through for now.

---

Yep, that's a hole in this PR. Will fix with what we discussed. Sorry, I should have done that before marking it as "ready." I got excited [and maybe a little ahead of myself] when the testsuite _finally_ passed 😄

---

One irritating thing about my implementation of this is that it takes a 3rd positional argument of `std::string &error_string` to return the error message, which, because this function is used for validation and parsing the extra options into a map, requires initializing a string in some places where we _know_ it won't be needed.

@trws's suggestion of using `std::expected<bool, std::string>` as the return type would probably be cleaner.

---

TIL about `std::expected`!

---

Yeah, a version of that is what I want us to do.  My ideal solution would be if we could use `std::expected<RetType, std::status_code>` but `status_code` isn't standard yet, and `error_code` is... easy to misuse.  We also don't require a C++23 stdlib yet.  The `boost::outcome` with the experimental header to use `boost::outcome<RetCode, boost::status_code>` is probably best, it's very hard to misuse and much faster to compile, but it requires just a bit of setup I haven't found time to do yet.

If you're up for it @wihobbs, it's not a lot of stuff and shouldn't be complicated and would really help out other places in sched too.  That said, I'd rather that be a separate PR from this since it's pretty orthogonal.

---

I opened #1330 for the task mentioned above.